### PR TITLE
perf: remove expensive reflection from xDS hot path

### PR DESCRIPTION
**Pull request by boxofrad**
### Description
Replaces the reflection-based implementation of proxycfg's `ConfigSnapshot.Clone` with code generated by [deep-copy](https://github.com/globusdigital/deep-copy).

While load testing server-based xDS (for consul-dataplane) we discovered this method is **extremely** expensive.

Screenshot 2022-10-06 at 12 16 42

The `ConfigSnapshot` struct, directly or indirectly, contains a copy of many of the structs in the `agent/structs` package, which creates a large graph for `copystructure.Copy` to traverse at runtime, on every proxy reconfiguration. This is particularly a problem when updating "global" resources such as wildcard intentions or `proxy-defaults` which triggers a thundering herd of updates (and in our tests caused Raft leadership instability).

### Testing & Reproduction steps
We test the completeness of `Clone` by using [gofuzz](https://github.com/google/gofuzz) to construct a "full" snapshot, cloning it, and then comparing the clone to the original.

The Envoy integration tests should also catch any regressions.

### CODE REVIEW

File: agent/proxycfg/deep-copy.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+
+# Uses: https://github.com/globusdigital/deep-copy
+deep-copy -pointer-receiver \
```
Comment (rboyer) [inline agent/proxycfg/deep-copy.sh:4]: Does this correctly nil out the cancelfunc maps?

File: agent/proxycfg/snapshot_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+	snapshot := new(ConfigSnapshot)
+	f.Fuzz(snapshot)
+
+	clone := snapshot.Clone()
+
+	diff := cmp.Diff(snapshot, clone,
+		cmpopts.IgnoreUnexported(indexedTarget{}),
+		cmpopts.IgnoreUnexported(pbpeering.PeeringTrustBundle{}),
+		cmpopts.IgnoreTypes(context.CancelFunc(nil)),
```
Comment (rboyer) [inline agent/proxycfg/snapshot_test.go:46]: The copied snapshots should _not_ have the cancelfunc maps populated IIRC.

File: proto/pbpeering/peering.deepcopy.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// generated by deep-copy -pointer-receiver -type PeeringTrustBundle -o ./peering.deepcopy.go ./; DO NOT EDIT.
+
+package pbpeering
+
+// DeepCopy generates a deep copy of *PeeringTrustBundle
```
Comment (rboyer) [inline proto/pbpeering/peering.deepcopy.go:5]: Shouldn't protobufs use `proto.Clone` instead?

File: proto/pbpeering/peering.go [hunk 4]
```diff
@@ -14,6 +15,7 @@ import (
 	"github.com/hashicorp/consul/agent/structs"
 	"github.com/hashicorp/consul/api"
 	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/consul/lib/stringslice"
```
Comment (rboyer) [inline proto/pbpeering/peering.go:18]: Stray import?

Comment (boxofrad) [reply proto/pbpeering/peering.go:18]: Yep! From my previous attempt at hand-rolling it, will remove.

File: GNUmakefile [hunk 5]
```diff
@@ -321,6 +322,10 @@ lint-tools:
 proto-tools:
 	@$(SHELL) $(CURDIR)/build-support/scripts/devtools.sh -protobuf
 
+.PHONY: codegen-tools
```
Comment (rboyer) [inline GNUmakefile:325]: I think we need a `make` target to correctly re-execute all of the codegen. Ideally also do something in CI that fails if a fresh run of the deep-copy generators produces a diff (like the `go.mod/go.sum` divergence checks).

Comment (boxofrad) [reply GNUmakefile:325]: Sounds good. Would something like this do the trick?

```
generate-deep-copy:
	go generate ./agent/structs ./proto/pbpeering ./agent/proxycfg
```

That'd technically run any other `go:generate` directives in those packages, so would it be better to run the `deep-copy.sh` scripts directly?

Comment (boxofrad) [reply GNUmakefile:325]: Added a make target and CI check ✅ (went with the latter approach)

**Review by rboyer — APPROVED** [review]

**Pull request merged**